Modify S2083&S6096(Education): Add Partial Path Traversal to pitfalls (#1243)
SonarSource/rspec: rules/S6096/java/how-to-fix-it/java-se.adoc

=== How does this work?

include::../../common/fix/how-does-this-work.adoc[]

=== Pitfalls

include::../../common/pitfalls/partial-path-traversal.adoc[]

For example, the following code is vulnerable to partial path injection. Note
that the string `targetDirectory` does not end with a path separator:

[source, java]
----

> static private String targetDirectory = "/Users/John";

Reviewer: Should this be also changed to uppercase 'TargetDirectory' like you did on the other example?

I have no clue on naming nomenclature so I will leave it to you to decide.

Author: No, I added the uppercase T because it is the C# naming convention IIRC

public void ExtractEntry(ZipFile zipFile) throws IOException {

    Enumeration<? extends ZipEntry> entries = zipFile.entries();
    ZipEntry entry = entries.nextElement();
    InputStream inputStream = zipFile.getInputStream(entry);

    File file = new File(entry.getName());

    String canonicalDestinationPath = file.getCanonicalPath();

    if (canonicalDestinationPath.startsWith(targetDirectory)) {
        Files.copy(inputStream, file.toPath(), StandardCopyOption.REPLACE_EXISTING, LinkOption.NOFOLLOW_LINKS);
    }
}
----

This check can be bypassed because `"/Users/Johnny".startsWith("/Users/John")`
returns `true`. Thus, for validation, `"/Users/John"` should actually be
`"/Users/John/"`.

> **Warning**: Some functions, such as `.getCanonicalPath`, remove the
> terminating path separator in their return value and cannot be used for
> validation. +

Reviewer: ```suggestion
**Warning**: Some functions, such as `.getCanonicalPath`, remove the
terminating path separator in their return value. +
```

getCanonicalPath only can not be used to retrieve the base path value (argument for startsWith), but it can still be used in the validation like you have on the compliant solution. Think this is too strong I would remove it in both.

Author: Agreed

> Validation code should be tested to ensure that it cannot be thwarted by this
> issue.

Reviewer: ```suggestion
Validation code should be tested to ensure that it cannot be exploited.
```

Author: If the issue is the word "thwarted" here, I would recommend the text below, because IMHO "exploited" can be misinterpreted, whereas talking about this issue in particular would make it clearer:

```suggestion
The validation code should be tested to ensure that it cannot be impacted by this
issue.
```
